### ISSUE TITLE
vagrant up assumes the wrong provider

### ISSUE THREAD

**Issue by mcandre** [hashicorp/vagrant#9291]
When developing with multiple providers, even for completely different boxes, `vagrant up` often gets mistaken about which provider to use, and will fail to find a box to spin up, even though it exists.

For example, if you have one VMware base box and one VirtualBox box, both locally imported:

```console
$ vagrant box list
mcandre/dilos (virtualbox, 0)
mcandre/macos (vmware_desktop, 0)
```

and a Vagrantfile that obviously points at the VirtualBox one:

```ruby
Vagrant.configure("2") do |config|
  config.vm.box = "mcandre/dilos"
end
```

Even though the mcandre/dilos box has exactly one provider (virtualbox), running `vagrant up` often complains that the box is missing.

I'm not sure which cache is causing this error, but for what it's worth, I have run `vagrant destroy -f; rm -rf .vagrant; vagrant up`, yielding the same erroneous behavior.

As a workaround, the provider can be explicitly configured (`vagrant up --provider virtualbox`), but it's quite stupid that Vagrant gets confused about this.

I'm using Vagrant 2.0.1 on macOS 10.13 High Sierra.

**Comment by mcandre** [comment]
Update:

This bug occurs with other boxes as well. I am unable to import the `debian/stretch64` box:

```console
$ vagrant up
Bringing machine 'default' up with 'vmware_fusion' provider...
==> default: Box 'debian/stretch64' could not be found. Attempting to find and install...
    default: Box Provider: vmware_desktop, vmware_fusion, vmware_workstation
    default: Box Version: >= 0
==> default: Loading metadata for box 'debian/stretch64'
    default: URL: https://vagrantcloud.com/debian/stretch64
The box you're attempting to add doesn't support the provider
you requested. Please find an alternate box or use an alternate
provider. Double-check your requested provider to verify you didn't
simply misspell it.

If you're adding a box from HashiCorp's Vagrant Cloud, make sure the box is
released.

Name: debian/stretch64
Address: https://vagrantcloud.com/debian/stretch64
Requested provider: ["vmware_desktop", "vmware_fusion", "vmware_workstation"]
```

**Comment by mcandre** [comment]
Specifically, the presence of the `vagrant-vmware-fusion` plugin blocks Vagrant from downloading VirtualBox boxes (???)

I'm working around this by temporarily uninstalling vagrant-vmware-fusion, when working with the debian/stretch64 VirtualBox box.

**Comment by jcocozzi** [comment]
I also experienced this issue and wrote to the google group, but received no response. 

https://groups.google.com/forum/#!searchin/vagrant-up/jeffrey.cocozziello%7Csort:date/vagrant-up/v49P6-V-mYc/-usCagRKBwAJ

**Comment by chrisroberts** [comment]
Determination of the provider to use happens prior to Vagrant reaching out to download a box from Vagrant Cloud. This is evident by running a simple `vagrant status` before a configured box has been installed. It will display the calculated provider that will be used. The default provider calculations happen here:

https://github.com/hashicorp/vagrant/blob/master/lib/vagrant/environment.rb#L311-L456

with a variety of ways to weight different providers. One simple way is to provide a `config.vm.provider :virtualbox` entry in the Vagrantfile. Modifying the calculated provider once we have reached the stage of talking with Vagrant Cloud is not something that is easily implemented and will likely have some unexpected implications.

If you would like to take a swing at dynamic modification of the provider based on the available boxes from Vagrant Cloud, I'll be happy to review a PR. Cheers!

**Comment by jcocozzi** [comment]
Hi Chris,
In my scenario (see the google group link), the vagrant cloud is not being used. Should I write up a new issue for this?

**Comment by chrisroberts** [comment]
Sure, that would be great.

Fri, Jan 05, 2018 at 05:00:17PM +0000 jcocozzi:

**Comment by martinanderssondotcom** [comment]
Why is this issue closed?

**Comment by General-Fault** [comment]
Ran into this issue with a box setup for hyperv. The problem occured when Powershell was not running under UAC. When "up"ing the hyper-v box, Vagrant would report that Virtualbox was not installed.
The cause was detected by calling "vagrant up --provider hyperv" to force use of the Hyper-V provider, whereby Vagrant then informed me that "The Hyper-V provider requires that Vagrant be run with administrative privileges." After switching to an admin Powershell window, I was able to provision the box.
